src/betterproto/__init__.py

```diff
@@ -343,20 +343,40 @@ def _pack_fmt(proto_type: str) -> str:
     }[proto_type]
 
 
+def dump_varint(value: int, stream: BinaryIO) -> None:
+    """Encodes a single varint and dumps it into the provided stream."""
+    if value < 0:
+        value += 1 << 64
+
+    bits = value & 0x7F
+    value >>= 7
+    while value:
+        stream.write((0x80 | bits).to_bytes())
+        bits = value & 0x7F
+        value >>= 7
+    stream.write(bits.to_bytes())
+
+
 def encode_varint(value: int) -> bytes:
     """Encodes a single varint value for serialization."""
-    b: List[int] = []
+    with BytesIO() as stream:
+        dump_varint(value, stream)
+        return stream.getvalue()
+
 
+def calculate_varint_size(value: int) -> int:
+    """Calculates the size in bytes that a value would take as a varint."""
     if value < 0:
         value += 1 << 64
```

Comment:
Will this mean that the value will always be 8 bytes?

```suggestion
    if value < 0:
        return 8
```

Reply:
Looked into this and from [this](https://protobuf.dev/programming-guides/encoding/#signed-ints) in the protobuf docs, it seems that this type of negative encoding should always return 10 bytes - which it does, until it hits anything less than `-(1<<63)` at which point the betterproto encoding breaks and starts looping back on itself and losing bytes and other oddities.

I've now added a limit to negative values beyond which these functions will return a `ValueError` to avoid confusion, and `calculate_varint_size` will now return `10` for all non-error negative values.